### ISSUE TITLE
global query filter - filter expression issue

### ISSUE THREAD

**Issue by dazinator** [aspnet/EntityFrameworkCore#11456]
I am trying to put together what I thought was a sane filter expression, but ef doesn't like it..

```
Exception message: 

> System.InvalidOperationException: The filter expression 'b => Property(b, value(Sample.EFCore.MultitenantDbContext`3+<>c__DisplayClass7_0`1[Sample.EFCore.SampleMultitenantDbContext,Sample.EFCore.Tenant,System.Guid,Sample.EFCore.Blog]).tenantIdPropertyName).Equals(Convert(value(Sample.EFCore.SampleMultitenantDbContext).TenantId, Object))' specified for entity type 'Blog' is invalid. The expression must accept a single parameter of type 'Sample.EFCore.Blog', return bool, and may not contain references to navigation properties.
  
Stack trace: 

> at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.ValidateQueryFilters(IModel model)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.Validate(IModel model)
   at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.Validate(IModel model)
   at Microsoft.EntityFrameworkCore.Internal.SqliteModelValidator.Validate(IModel model)
```

As far as I can tell my filter expression is compliant:

"The expression must accept a single parameter of type 'Sample.EFCore.Blog', return bool, and may not contain references to navigation properties."

Here is my expression:

```
   modelBuilder.Entity<T>().HasQueryFilter(b => EF.Property<TIdType>(b, tenantIdPropertyName).Equals(TenantId));
```

Which complies with:
- [x] accept a single parameter of type 'Sample.EFCore.Blog'
- [x] return bool
- [x] may not contain references to navigation properties.

### Steps to reproduce

Create a base dbcontext:

```csharp
   public abstract class MultitenantDbContext<TDbContext, TTenant, TIdType> : DbContext
        where TDbContext : DbContext
        where TTenant : class
    {
        private readonly Task<TTenant> _tenant;
        private Lazy<TIdType> _tenantId;

        private TIdType TenantId
        {
            get
            {
                return _tenantId.Value;
            }
        }

        private static List<Action> _setTenantIdOnSaveCallbacks = new List<Action>();

        public MultitenantDbContext(DbContextOptions<TDbContext> options, Task<TTenant> tenant) : base(options)
        {
            _tenant = tenant;
            _tenantId = new Lazy<TIdType>(() =>
            {
                var t = _tenant.Result;
                return GetTenantId(t);
            });
        }

        protected abstract TIdType GetTenantId(TTenant tenant);

        protected void HasTenantIdFilter<T>(ModelBuilder modelBuilder, string tenantIdPropertyName)
          where T : class
        {
            modelBuilder.Entity<T>().Property<TIdType>(tenantIdPropertyName);
            modelBuilder.Entity<T>().HasQueryFilter(b => EF.Property<TIdType>(b, tenantIdPropertyName).Equals(TenantId));

            Action action = () =>
            {
                SetTenantIdProperty<T>(tenantIdPropertyName);
            };
            _setTenantIdOnSaveCallbacks.Add(action);

        }

        public override int SaveChanges(bool acceptAllChangesOnSuccess)
        {
            SetTenantIdOnSave();
            return base.SaveChanges(acceptAllChangesOnSuccess);
        }

        public override int SaveChanges()
        {
            SetTenantIdOnSave();
            return base.SaveChanges();
        }

        public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken = default(CancellationToken))
        {
            SetTenantIdOnSave();
            return await base.SaveChangesAsync(cancellationToken);
        }

        public override async Task<int> SaveChangesAsync(bool acceptAllChangesOnSuccess, CancellationToken cancellationToken = default(CancellationToken))
        {
            SetTenantIdOnSave();
            return await base.SaveChangesAsync(acceptAllChangesOnSuccess, cancellationToken);
        }

        private void SetTenantIdProperty<TEntity>(string propertyName)
            where TEntity : class
        {
            foreach (var item in ChangeTracker.Entries<TEntity>())
            {
                if (item.State == EntityState.Added)
                {
                    item.Property(propertyName).CurrentValue = TenantId;
                }
            }
        }

        private void SetTenantIdOnSave()
        {
            ChangeTracker.DetectChanges();
            foreach (var item in _setTenantIdOnSaveCallbacks)
            {
                item();
            }
        }
    }
```

and a Tenant class to use in a moment:

```csharp
    public class Tenant
    {
        public Tenant(Guid tenantGuid, string name)
        {
            TenantGuid = tenantGuid;
            Name = name;
        }

        public Guid TenantGuid { get; set; }
        public string Name { get; set; }
    }

```

Now derive a DbContext from the base:

```csharp

  public class SampleMultitenantDbContext : MultitenantDbContext<SampleMultitenantDbContext, Tenant, Guid>
    {

        private const string TenantIdPropertyName = "TenantId";

        public SampleMultitenantDbContext(DbContextOptions<SampleMultitenantDbContext> options, Task<Tenant> tenant) : base(options, tenant)
        {
        }

        public DbSet<Blog> Blogs { get; set; }

        protected override Guid GetTenantId(Tenant tenant)
        {
            return tenant.TenantGuid;
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            HasTenantIdFilter<Blog>(modelBuilder, TenantIdPropertyName);
        }      
    }

```

Now use the `SampleMultitenantDbContext` to see the error:

```csharp

// in configure services:
services.AddSingleton<Task<Tenant>>(Task.Run<Tenant>(()=>new Tenant(Guid.NewGuid(), "foo")));
services.AddDbContext<SampleMultitenantDbContext>((options) =>
            {
                var path = System.Environment.GetFolderPath(System.Environment.SpecialFolder.Personal);
                var dbFilePath = System.IO.Path.Combine(path, "DotnettencyMultitenantDb.db");
                options.UseSqlite($"Filename={dbFilePath};");
            });



// in Configure()

            app.Map("/Blogs", (b) =>
            {
                b.Run(async (context) =>
                {
                    var tenantAwaredDbContext = context.RequestServices.GetRequiredService<SampleMultitenantDbContext>();
                    var blogs = await tenantAwaredDbContext.Blogs.ToArrayAsync();
                    foreach (var item in blogs)
                    {
                        await context.Response.WriteAsync($"Blog Id: {item.BlogId}, Url: {item.Url}");
                    }
                });
            });

```

### Further technical details
EF Core version: (found in project.csproj or packages.config)
 <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.0" />
Database Provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
sqlite
Operating system: 
Visual Studio 2017 latest stable release

**Comment by smitpatel** [comment]
This reproduces in 2.0.1 packages. But it is working in 2.1.0-preview1-final package. Closing as fixed. You can try out preview1 package which are available on Nuget.

**Comment by dazinator** [comment]
Does this mean that in order to get this working for an asp.net core 2.0.0 website, we will have to upgrade to asp.net core 2.1.0? Or can EF 2.1.0 be consumed with an asp.net core 2.0.0 website - i.e are the versions of asp.net core and ef tightly coupled?

**Comment by smitpatel** [comment]
@divega - Can you answer above question?

**Comment by divega** [comment]
@dazinator 
TL;DR: it is recommended that you use EF Core 2.1 with ASP.NET Core 2.1. Although most things may work with ASP.NET Core 2.0, we don't test this scenario as exhaustively so you may run into unexpected issues.

In more detail, EF Core 2.1 is *just* a .NET Standard 2.0 library, so in theory it should work with any platform that implements it. In practice: 
- So far, EF Core 2.1 is known to generally work with .NET Core 2.0, however
- We only do exhaustive testing of EF Core 2.1 on .NET Core 2.1 and .NET Framework and therefore those are the recommended combinations
- If anyone reports an issue using EF Core 2.1 on .NET Core 2.0 we will take a look and may decide not to fix it depending on the impact, workarounds, etc.

When you use EF Core 2.1 with ASP.NET Core 2.0 on .NET Core 2.0 there is a bit of additional complexity involved because:
- The ASP.NET Core meta-package (used by default on ASP.NET Core 2.0 templates) already includes the 2.0 EF Core bits, so effectively you are lifting the versions of the EF Core packages and all its dependencies. 
- EF Core and ASP.NET Core have lost of common dependencies, e.g. logging, dependency injection, in-memory caching, etc. So now you are using ASP.NET Core 2.0 with newer versions of those. While in theory that should be ok (we have a very high bar for breaking changes in minor releases), in practice we don't test all those combinations.

**Comment by dazinator** [comment]
Thanks for the explaination. So if I understand correctly, it is the shared dependencies that cause the coupling issue with asp.net core version. So if, for example, EF had zero dependencies, then in theory it wouldnt matter about what version of asp.net you were on. That would be an incredibly nice and advantageous place for a consumer of EF tobe because I could then adopt new EF releases without being forced with the sometimes mammoth and risky business of upgrading an entire asp.net core application. Is there an ILMerge equivalent for dot net core yet?  :-) seriously is merging dependencies something you would consider? I guess I could fork the repo and ILMerge the desktop version (we actually run on .net462).

Sorry if this comes across as a bit blunt but hopefully you will forgive me for asking, given that this bug replicates in 2.0.1, why can't there be a patch release?

**Comment by divega** [comment]
> Sorry if this comes across as a bit blunt but hopefully you will forgive me for asking, given that this bug replicates in 2.0.1, why can't there be a patch release?

No worries, it is a fair question. First of all, the limitation is in a model validation that was supposed to prevent query filters from using navigation properties. That code assumed that you would pass the property name to EF.Property as a constant. 

A few reasons we haven't tried to fix this in a patch are:
- It is not a regression, i.e. the limitation existed when the feature was added in 2.0.0 
- We did not hear from any customers hitting this when we released. I suspect that correlates to the fact that the limitation doesn't manifest in simple usage. It only manifest in you repro because in your usage the property name ends up being a parameter that gets captured in the expression. 
- Even if we had seen customers hitting it, there seem to be multiple ways to workaround it, e.g. you could refactor your code to capture EF.Property(TenantIdPropertyName) in SampleMultitenantDbContext, or build it into the expression as a constant in MultitenantDbContext. 

Now that this validation was removed in 2.1, the problem just doesn't exist. 

> being able to adopt new EF releases without being forced with the sometimes mammoth and risky business of upgrading the entire asp.net core application would be very nice.

Fair feedback as well. The way we build EF Core right now it only makes sense to release it in lockstep with ASP.NET Core and the rest of the .NET Core platform, but that may change in the future. The feedback is also interesting for me because you seem to imply that the risk of upgrading from ASP.NET Core 2.0 to 2.1 is much higher than upgrading only EF Core. From the perspective of EF Core being just a smaller slice of your application's dependencies, I agree. But the policy ASP.NET Core and EF Core regarding breaking changes in minor releases is the same.

>  I guess I could fork the repo and have a go at ILMerging the desktop version

We haven't considered doing this. If you are willing to do something like this, I believe it may be more less hassle to just mix 2.0 and 2.1 components and test that it works for you, until you are ready to jump to 2.1.

**Comment by dazinator** [comment]
Thanks for the details answer, that makes sense.

> there seem to be multiple ways to workaround it

I'd love to find a workaround for this, could you help shed some light on what I am not seeing..? I am sure I am being blind.

> e.g. you could refactor your code to capture EF.Property(TenantIdPropertyName) in SampleMultitenantDbContext

Not sure what you mean? There is no `EF.Property(TenantIdPropertyName)` method, only `EF.Property(entityInstance, TenantIdPropertyName)

The only place I have access to the entity instance is within the query filter expression.. which is why I am doing this:

```
modelBuilder.Entity<T>().HasQueryFilter(b => EF.Property<TIdType>(b, tenantIdPropertyName).Equals(TenantId));
```
      
> or build it into the expression as a constant in MultitenantDbContext.

I can't build it in as a constant, because the current tenant id is not a constant. The current tenant ID is determined by certain properties of the current HttpRequest. I treat it as a `Lazy<TIdKind>`.

Basically, I am trying to provide a library that others can use to derive a multi-tenant dbcontext very easily, and I would like it to support asp.net core 2.0 applications. So upgrading to 2.1.0 to get this fix isn't my preferred solution if any other workaround is possible!

**Comment by dazinator** [comment]
Spoke to soon..

![image](https://user-images.githubusercontent.com/3176632/39379540-7ec06434-4a53-11e8-9041-84237eb6cc9a.png)

**Comment by dazinator** [comment]
Ok in the end I managed to solve this by adapting the expression and some help from a kind soul on stack overflow.



In MultitenantDbContext:
```

  protected void HasTenantIdFilter<T>(ModelBuilder modelBuilder, string tenantIdPropertyName, Expression<Func<T, TIdType>> idExpression)
          where T : class
        {

            modelBuilder.Entity<T>().Property<TIdType>(tenantIdPropertyName);

            var newExp = Expression.Lambda<Func<T, bool>>(
                    Expression.Equal(idExpression.Body, Expression.Property(Expression.Constant(this),
                    typeof(MultitenantDbContext<TDbContext, TTenant, TIdType>),
                    nameof(TenantId))), idExpression.Parameters);

            modelBuilder.Entity<T>().HasQueryFilter(newExp);

            Action<DbContext> action = (db) =>
            {
                SetTenantIdProperty<T>(tenantIdPropertyName, TenantId, db);
            };
            _setTenantIdOnSaveCallbacks.Add(action);

        }


```

In SampleDbContext:

```

 protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            HasTenantIdFilter<Blog>(modelBuilder, TenantIdPropertyName, (b) => EF.Property<Guid>(b, TenantIdPropertyName));
        }

```

**Comment by dazinator** [comment]
> The feedback is also interesting for me because you seem to imply that the risk of upgrading from ASP.NET Core 2.0 to 2.1 is much higher than upgrading only EF Core.

I know this is old now, but you are right, I did imply this :-)
That's because I have lived with the track record of asp.net core upgrades from RC preview through to 2.0, in which breaking changes have been commonplace. For example if you read https://docs.microsoft.com/en-us/aspnet/core/migration/1x-to-2x/?view=aspnetcore-2.1 you realise that that it's not just project or code changes that need to be made during upgrades, build tooling also changes, and even the build tooling changing introduces breaking changes that were often unannounced.. I understand that 2.0 to 2.1 _should be a lot less painful_ but the reality is often different. The risk is greater because the surface area of upgrading an app, instead of a single dependency is greater.
